nilearn/_utils/data_gen.py

```diff
@@ -866,93 +866,228 @@ def create_fake_bids_dataset(base_dir='',
+                                                        rand_gen,
+                                                        with_confounds,
+                                                        confounds_tag)
+                        elif entities[0] in ENTITIES:
+                            entity = entities[0]
+                            labels = entities[1]
+                            for i_label in labels:
+                                fields = [subject,
+                                          session,
+                                          'task-' + task,
+                                          f"{entity}-{i_label}"]
+                                _write_bids_derivative_func(func_path,
+                                                            _file_id(fields,
+                                                                     n_run,
+                                                                     run),
+                                                            N_VOXELS,
+                                                            rand_gen,
+                                                            with_confounds,
+                                                            confounds_tag)
     return 'bids_dataset'
 
 
+def _file_id(fields: list[str], n_run: int, run: str) -> str:
+    """Create BIDS filename from list of entity-label pairs.
+
+    Parameters
+    ----------
+    fields : list[str]
+        _description_
+    n_run : int
+        _description_
+    run : str
+        _description_
+
+    Returns
+    -------
+    BIDS filename : str
+        'file_id'.
+    """
+    if '' in fields:
+        fields.remove('')
+    file_id = '_'.join(fields)
+    if n_run > 1:
+        file_id += f'_{run}'
+    return file_id
+
+
+def _write_bids_raw_func(func_path: str,
+                         file_id: str,
+                         n_voxels: int,
+                         rand_gen: np.random.RandomState) -> None:
+    """Create BIDS functional raw and events files.
+
+    Parameters
+    ----------
+    func_path : str
+        _description_
+    file_id : str
+        _description_
+    n_voxels : int
+        _description_
+    rand_gen : np.random.RandomState
+        _description_
+    """
+    REPETITION_TIME = 1.5
+    N_TIME_POINTS = 100
+
+    bold_path = os.path.join(func_path, f'{file_id}_bold.nii.gz')
+    write_fake_bold_img(bold_path,
+                        [n_voxels, n_voxels, n_voxels, N_TIME_POINTS],
+                        random_state=rand_gen)
+
+    events_path = os.path.join(func_path, f'{file_id}_events.tsv')
+    basic_paradigm().to_csv(events_path, sep='\t', index=None)
+
+    param_path = os.path.join(func_path, f'{file_id}_bold.json')
+    with open(param_path, 'w') as param_file:
+        json.dump({'RepetitionTime': REPETITION_TIME},
+                  param_file)
+
+
+def _write_bids_derivative_func(func_path: str,
```

Comment:
@jeromedockes 

Related to our conversation regarding adding type annotation.

One possible way to start doing it bit by bit is by adding type annotation when extracting functions during refactoring. 

Doing it here to show how this could look like.

Reply:
Will open a separate issue to keep track of type annotation related conversations.